Automattic/wp-calypso
# Store: Add query arg support to fetching product categories

This PR updates the Product Category fetching code to support the sending of query arguments with a request. Previously, the product layer code could only make one request for the first page of results. This will be used as a part of #20295, so that I can get multiple pages of responses, use the search argument, request more than the API default of 10 items, and get additional information back like the total number of categories on the site.

The README probably explains the changes best if you want to start with that. The short version is: The state structure and selectors have been updated to be patterned off of what we are doing in other places like orders, products, and product reviews. (I.e. storing all category data in `productCategories.items` instead of just `productCategories`, tracking them under `queries` etc).

Sorry that this PR is a bit bigger, but most of the lines come from README/documentation and tests.
Since this is a larger PR though, we should probably have a couple sets of eyes review it.

To Test:
* Make sure all tests pass: `npm run test-client client/extensions/woocommerce/state/`
* Visit `http://calypso.localhost:3000/store/promotion/:site` and select `Product Categories` under the apples to dialog. Make sure your categories still load
* Visit `http://calypso.localhost:3000/store/product/:site` and make sure categories still show up in the category autocomplete.
* (note that you will only see 10 categories before and after this PR is applied. That is logged at #20339).
